src/perf_mainloop.cc

```diff
@@ -142,52 +143,44 @@ static DDRes signalfd_setup(pollfd *pfd) {
 static inline DDRes worker_process_ring_buffers(PEvent *pes, int pe_len,
                                                 DDProfContext *ctx,
                                                 int64_t *now_ns,
-                                                bool *timeout) {
+                                                int *ring_buffer_start_idx) {
   // While there are events to process, iterate through them
   // while limiting time spent in loop to at most PSAMPLE_DEFAULT_WAKEUP_MS
   int64_t loop_start_ns = now_nanos();
   int64_t local_now_ns;
 
+  int start_idx = *ring_buffer_start_idx;
   bool events;
   do {
     events = false;
-    for (int i = 0; i < pe_len; ++i) {
-      RingBuffer *rb = &pes[i].rb;
-
-      // Memory-ordering safe access of ringbuffer elements:
-      // https://github.com/torvalds/linux/blob/v5.16/tools/include/linux/ring_buffer.h#L59
-      uint64_t head = __atomic_load_n(&rb->region->data_head, __ATOMIC_ACQUIRE);
-      uint64_t tail = rb->region->data_tail;
-      while (tail != head) {
-        events = true;
-
-        // Attempt to dispatch the event
-        struct perf_event_header *hdr = rb_seek(rb, tail);
-        assert(hdr->size > 0);
+    for (int i = start_idx; i < pe_len; ++i) {
+      ddprof::PerfRingBufferReader reader(pes[i].rb);
+
+      ddprof::ConstBuffer buffer = reader.read_all_available();
+      while (!buffer.empty()) {
+        auto *hdr = reinterpret_cast<const perf_event_header *>(buffer.data());
         DDRes res = ddprof_worker_process_event(hdr, pes[i].watcher_pos, ctx);
 
-        // We've processed the current event, so we can advance the ringbuffer
-        tail += hdr->size;
+        // \fixme{nsavoire} free slot as soon as possible ?
+        // reader.advance(hdr->size);
```

Comment:
I'm missing why would we not advance here ?

Reply:
For the moment, I am keeping the same behaviour as before, advance is done once all available samples of the current ring buffer are processed (it's done by `reader`'s destructor).
Since unwinding is not that fast, it could be advantageous to advance for each sample, doing more writes to the ring buffer reader positions but freeing space faster for the writer.

Reply:
Agreed, I think this will make things better 👍